torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -510,3 +512,38 @@ def shard_parameter(
 
     # Now we can set the attribute appropriately.
     setattr(module, param_name, st)
+
+def custom_sharded_op(op):
+    """
+    Provides a way for users to write their own custom sharded operator. This
+    can be used to override existing ShardedTensor operatorss or write a new
+    one not supported by ShardedTensor. If the operator in question is covered
+    by ``__torch_function__`` dispatch and has a ShardedTensor as any of its
+    parameters, the function provided will be invoked for that operator.
+
+    Example::
+        >>> @custom_sharded_op(torch.nn.functional.linear)
+        >>> def my_custom_sharded_linear(types, args, kwargs, process_group):
+        >>>   ....
+        >>>
+        >>> input = torch.rand(10, 32)
+        >>> weight = _sharded_tensor.rand(32, 16)
+        >>> bias = torch.rand(16)
+        >>> # This will call 'my_custom_sharded_linear'
+        >>> torch.nn.functional.linear(input, weight, bias)
+
+    The types, args and kwargs parameters are the same parameters that are
```

Comment:
I just take a look on all the implementations we have, the `types` parameter seems not getting used anywhere, I am wondering if we should get rid of this argument both for custom op and for our own ops, `__torch_function__` can have it, but it seems like we don't need it for the actual computation?

Reply:
I guess its safer to preserve types as it is provided to the user in `__torch_function__` as well.